File: python/quorum/crypto.py
"""quorum.crypto — Argon2id share derivation and webhook signature verification.

Argon2id parameters match quorum-engine/src/crypto/kdf.rs exactly:
    m_cost   = 65_536 KiB  (64 MB)
    t_cost   = 3
    p_cost   = 1
    hash_len = 32 bytes
"""
from __future__ import annotations

import hashlib
import hmac
import secrets
from typing import Union

from argon2.low_level import hash_secret_raw, Type

# ── Constants — must match quorum-engine kdf.rs ───────────────────────────────
_M_COST    = 65_536   # KiB
_T_COST    = 3
_P_COST    = 1
_HASH_LEN  = 32
_SALT_LEN  = 32
_MIN_PIN   = 6


def derive_share(pin: str, salt_hex: str) -> str:
    """Derive an Argon2id key share from a PIN and hex-encoded salt.

    This is the server-side path for USSD members or server-to-server
    share computation. Browser members use the JS SDK instead.

    Args:
        pin:      The member's PIN or passphrase (min 6 characters).
        salt_hex: The member's unique 32-byte salt, hex-encoded (64 chars).

    Returns:
        The 32-byte share as a lowercase hex string.

    Raises:
        ValueError: If the PIN is too short or the salt is the wrong length.

    Example::

        share = derive_share("284751", "a3f2e1d4..." * 2)
        # submit to POST /v1/authorize as share_hex
    """
    if len(pin) < _MIN_PIN:
        raise ValueError(f"PIN must be at least {_MIN_PIN} characters")

    if len(salt_hex) != _SALT_LEN * 2:
        raise ValueError(f"salt_hex must be {_SALT_LEN * 2} hex characters")

    try:
        salt = bytes.fromhex(salt_hex)
    except ValueError as exc:
        raise ValueError(f"salt_hex is not valid hex: {exc}") from exc

    raw = hash_secret_raw(
        secret=pin.encode(),
        salt=salt,
        time_cost=_T_COST,
        memory_cost=_M_COST,
        parallelism=_P_COST,
        hash_len=_HASH_LEN,
        type=Type.ID,
    )
    return raw.hex()


def verify_webhook_signature(
    payload:   Union[bytes, str],
    secret:    str,
    signature: str,
) -> bool:
    """Verify a Quorum webhook HMAC-SHA256 signature.

    Quorum signs webhook payloads with HMAC-SHA256. The signature arrives in
    the ``X-Quorum-Signature-256`` header as ``sha256=<hex>``.

    Uses ``hmac.compare_digest`` for constant-time comparison.

    Args:
        payload:   The raw request body (bytes or str).
        secret:    The webhook signing secret provided at registration.
        signature: The full value of the X-Quorum-Signature-256 header.

    Returns:
        True if the signature is valid, False otherwise.

    Example (Flask)::

        from flask import request
        from quorum import verify_webhook_signature

        @app.route("/webhooks/quorum", methods=["POST"])
        def handle():
            sig = request.headers.get("X-Quorum-Signature-256", "")
            if not verify_webhook_signature(request.data, WEBHOOK_SECRET, sig):
                return "Unauthorized", 401
            event = request.json
            ...
    """
    if isinstance(payload, str):
        payload = payload.encode()

    mac = hmac.new(secret.encode(), payload, hashlib.sha256)
    expected = f"sha256={mac.hexdigest()}"
    return hmac.compare_digest(expected, signature)


def generate_salt() -> str:
    """Generate a cryptographically random 32-byte salt, hex-encoded.

    Used during group enrolment ceremonies. Each member gets a unique salt.

    Returns:
        64-character lowercase hex string.
    """
    return secrets.token_hex(_SALT_LEN)

File: js/src/crypto.ts
// src/crypto.ts — Argon2id share derivation and HMAC-SHA256 webhook verification
//
// # Share derivation
// deriveShare() runs Argon2id with the platform parameters enforced by the
// engine. It works in both browsers (via argon2-browser WebAssembly) and
// Node.js (via @node-rs/argon2 if available, fallback to argon2-browser).
//
// # Argon2id parameters
// These MUST match the engine constants in quorum-engine/src/crypto/kdf.rs:
//   M_COST = 65_536 KiB   (64 MB)
//   T_COST = 3
//   P_COST = 1
//   OUTPUT = 32 bytes

const ARGON2_M_COST = 65536; // 64 MB
const ARGON2_T_COST = 3;
const ARGON2_P_COST = 1;
const ARGON2_HASH_LEN = 32;
const MIN_PIN_LEN = 6;
const SALT_LEN = 32;

// ── Argon2id share derivation ────────────────────────────────────────────────

/**
 * Derive an Argon2id key share from a PIN and hex-encoded salt.
 *
 * @param pin     - The member's PIN or passphrase (min 6 characters)
 * @param saltHex - The member's unique 32-byte salt, hex-encoded (64 hex chars)
 * @returns       - The 32-byte share as a lowercase hex string
 *
 * @example
 * const share = await deriveShare('284751', 'a3f2e1d4....');
 * // submit share to POST /v1/authorize
 */
export async function deriveShare(pin: string, saltHex: string): Promise<string> {
  if (pin.length < MIN_PIN_LEN) {
    throw new Error(`PIN must be at least ${MIN_PIN_LEN} characters`);
  }
  if (saltHex.length !== SALT_LEN * 2) {
    throw new Error(`salt_hex must be ${SALT_LEN * 2} hex characters (${SALT_LEN} bytes)`);
  }

  const salt = hexToBytes(saltHex);
  const pinBytes = new TextEncoder().encode(pin);

  // Try Node.js native path first (server-side SDK usage)
  if (typeof process !== 'undefined' && process.versions?.node) {
    return deriveShareNode(pinBytes, salt);
  }

  // Browser path: argon2-browser WebAssembly
  return deriveShareBrowser(pinBytes, salt);
}

async function deriveShareBrowser(pin: Uint8Array, salt: Uint8Array): Promise<string> {
  // Dynamic import so the WASM only loads when needed
  const { hash, ArgonType } = await import('argon2-browser');

  const result = await hash({
    pass: pin,
    salt,
    type: ArgonType.Argon2id,
    mem: ARGON2_M_COST,
    time: ARGON2_T_COST,
    parallelism: ARGON2_P_COST,
    hashLen: ARGON2_HASH_LEN,
  });

  return bytesToHex(result.hash);
}

async function deriveShareNode(pin: Uint8Array, salt: Uint8Array): Promise<string> {
  // Node.js: use @node-rs/argon2 if available (native bindings, ~10x faster)
  // Falls back to argon2-browser WASM if not installed.
  try {
    const { hash } = await import('@node-rs/argon2');
    const result = await hash(pin, {
      salt,
      algorithm: 2, // Argon2id
      memoryCost: ARGON2_M_COST,
      timeCost: ARGON2_T_COST,
      parallelism: ARGON2_P_COST,
      outputLen: ARGON2_HASH_LEN,
    });
    // @node-rs/argon2 returns the raw hash as a Buffer
    return bytesToHex(typeof result === 'string' ? hexToBytes(result) : new Uint8Array(result));
  } catch {
    // Fallback to browser WASM
    return deriveShareBrowser(pin, salt);
  }
}

// ── Webhook signature verification ───────────────────────────────────────────

/**
 * Verify a Quorum webhook signature.
 *
 * Quorum signs webhook payloads with HMAC-SHA256. The signature is sent in
 * the `X-Quorum-Signature-256` header as `sha256=<hex>`.
 *
 * @param payload   - The raw request body bytes
 * @param secret    - The webhook signing secret set at registration
 * @param signature - The full value of the X-Quorum-Signature-256 header
 * @returns         - true if the signature is valid
 *
 * @example
 * const body = await request.text();
 * const sig  = request.headers.get('X-Quorum-Signature-256');
 * const ok   = await verifyWebhookSignature(
 *   new TextEncoder().encode(body), process.env.WEBHOOK_SECRET!, sig!
 * );
 */
export async function verifyWebhookSignature(
  payload:   Uint8Array | string,
  secret:    string,
  signature: string,
): Promise<boolean> {
  const data = typeof payload === 'string'
    ? new TextEncoder().encode(payload)
    : new Uint8Array(payload);

  const keyMaterial = await crypto.subtle.importKey(
    'raw',
    new TextEncoder().encode(secret),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign'],
  );

  const mac = await crypto.subtle.sign('HMAC', keyMaterial, data);
  const expected = 'sha256=' + bytesToHex(new Uint8Array(mac));

  // Constant-time comparison
  return timingSafeEqual(expected, signature);
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function hexToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substring(i, i + 2), 16);
  }
  return bytes;
}

function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
}

/** Timing-safe string comparison to prevent timing attacks on HMAC checks. */
function timingSafeEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return diff === 0;
}
